File: game/src/main/java/com/exjets/aetheria/core/Equipment.java
package com.exjets.aetheria.core;

import java.util.EnumMap;
import java.util.Map;

/** Worn gear, one item per slot, contributing bonuses to the combat formulas. */
public class Equipment {

    private final Map<EquipSlot, String> worn = new EnumMap<>(EquipSlot.class);

    public String get(EquipSlot slot) {
        return worn.get(slot);
    }

    public boolean isWearing(String itemId) {
        return worn.containsValue(itemId);
    }

    /** @return the item that was previously in the slot, or null */
    public String equip(String itemId) {
        ItemDef def = ItemRegistry.get(itemId);
        if (!def.equipable()) {
            throw new IllegalArgumentException(def.name() + " cannot be equipped");
        }
        return worn.put(def.slot(), itemId);
    }

    public String unequip(EquipSlot slot) {
        return worn.remove(slot);
    }

    public Map<EquipSlot, String> all() {
        return worn;
    }

    public int attackBonus() {
        return sum(ItemDef::attackBonus);
    }

    public int strengthBonus() {
        return sum(ItemDef::strengthBonus);
    }

    public int defenceBonus() {
        return sum(ItemDef::defenceBonus);
    }

    public void clear() {
        worn.clear();
    }

    private int sum(java.util.function.ToIntFunction<ItemDef> field) {
        int total = 0;
        for (String id : worn.values()) {
            total += field.applyAsInt(ItemRegistry.get(id));
        }
        return total;
    }
}

File: src/game/core/equipment.ts
import { EquipSlot, ItemDef, itemDef } from "./items";

/** Worn gear, one item per slot, feeding the combat formulas. */
export class Equipment {
  private readonly worn = new Map<EquipSlot, string>();

  get(slot: EquipSlot): string | null {
    return this.worn.get(slot) ?? null;
  }

  /** @returns the item that was in the slot before, or null */
  equip(id: string): string | null {
    const def = itemDef(id);
    if (!def.slot) throw new Error(`${def.name} cannot be equipped`);
    const previous = this.worn.get(def.slot) ?? null;
    this.worn.set(def.slot, id);
    return previous;
  }

  unequip(slot: EquipSlot): string | null {
    const removed = this.worn.get(slot) ?? null;
    this.worn.delete(slot);
    return removed;
  }

  entries(): [EquipSlot, string][] {
    return [...this.worn.entries()];
  }

  attackBonus(): number {
    return this.sum((def) => def.attackBonus);
  }

  strengthBonus(): number {
    return this.sum((def) => def.strengthBonus);
  }

  defenceBonus(): number {
    return this.sum((def) => def.defenceBonus);
  }

  clear(): void {
    this.worn.clear();
  }

  private sum(field: (def: ItemDef) => number): number {
    let total = 0;
    for (const id of this.worn.values()) total += field(itemDef(id));
    return total;
  }
}
